File: NeetCode/Python/Arrays/Easy/205.IsomorphicStrings.py
# Given two strings s and t, determine if they are isomorphic.
# Two strings s and t are isomorphic 
# if the characters in s can be replaced to get t.

def isIsomorphic(s, t):
    s_dict, t_dict = {}, {}

    for i, j in zip(s, t):
        if (i in s_dict and s_dict[i] != j) or (j in t_dict and t_dict[j] != i):
            return False

        s_dict[i] = j
        t_dict[j] = i

    return True

print(isIsomorphic("egg", "add"))
print(isIsomorphic("foo", "bar"))
print(isIsomorphic("paper", "title"))
print(isIsomorphic("badc", "baba"))

File: NeetCode/Swift/Arrays/Easy/205.IsomorphicStrings.swift
//Given two strings s and t, determine if they are isomorphic.
//Two strings s and t are isomorphic 
//if the characters in s can be replaced to get t.

func isIsomorphic(_ s: String, _ t: String) -> Bool {
    //a letter of a String is a Character
    //Characters can only have 1 element
    var sDict = [Character: Character]()
    var tDict = [Character: Character]()

    //using subscripts returns a Character Type
    //i and j are characters
    for (i, j) in zip(s, t) {
        if (sDict[i] != nil && sDict[i] != j) || (tDict[j] != nil && tDict[j] != i) {
            return false
        }

        sDict[i] = j 
        tDict[j] = i
    }

    return true
}

print(isIsomorphic("egg", "add"))
print(isIsomorphic("foo", "bar"))
print(isIsomorphic("paper", "title"))
print(isIsomorphic("badc", "baba"))